wa/workloads/speedometer/__init__.py

```diff
@@ -12,57 +12,221 @@
+
+        logging.info(
+            "Using port {} for temporary webserver".format(self.webserver_port)
+        )
+
+        self.speedometer_url = "http://localhost:{}/Speedometer2.0/index.html".format(self.webserver_port)
+
+        # Host a copy of Speedometer locally
+        subprocess.check_output(
+            "tar -xf {} -C {}".format(self.archive_asset, context.output_directory),
+            shell=True,
+        )
+
+        webserver_cmd = "cd {} && python3 -mhttp.server {}".format(
+            self.document_root, self.webserver_port
+        )
+
+        with open(os.devnull, "w") as dev_null:
+            self.webserver = subprocess.Popen(
+                webserver_cmd, shell=True, stdout=dev_null, stderr=dev_null
+            )
+
+        subprocess.check_output(
+            "adb reverse tcp:{0} tcp:{0}".format(self.webserver_port), shell=True
+        )
+
+        # Clear the cache...
+        self.target.execute("pm clear {}".format(self.chrome_package, as_root=True))
+
+        # Launch the browser for the first time...
+        browser_launch_cmd = "am start -a android.intent.action.VIEW -d {} {}".format(
+            self.speedometer_url, self.chrome_package
+        )
+
+        self.target.execute(browser_launch_cmd)
+        time.sleep(1)
+        self.target.execute("am force-stop {}".format(self.chrome_package))
+        time.sleep(1)
+
+        # Pull the preferences file from the device, modify it, and push it back.
+        # This is done to bypass the 'first launch' screen of the browser.
+        self.preferences_xml = "{}_preferences.xml".format(self.chrome_package)
+
+        self.target.pull(
+            "/data/data/{}/shared_prefs/{}".format(
+                self.chrome_package, self.preferences_xml
+            ),
+            context.output_directory,
+            as_root=True,
+        )
+
+        with open(
+            os.path.join(context.output_directory, self.preferences_xml)
+        ) as read_fh:
+            lines = read_fh.readlines()
+            lines.insert(
+                len(lines) - 1, '<boolean name="first_run_flow" value="true" />\n'
+            )
+            lines.insert(
+                len(lines) - 1,
+                '<boolean name="first_run_tos_accepted" value="true" />\n',
+            )
+            lines.insert(
+                len(lines) - 1,
+                '<boolean name="first_run_signin_complete" value="true" />\n',
+            )
+            with open(
+                os.path.join(context.output_directory, self.preferences_xml + ".new"),
+                "w",
+            ) as write_fh:
+                for line in lines:
+                    write_fh.write(line)
+
+        self.target.push(
+            os.path.join(context.output_directory, self.preferences_xml + ".new"),
+            "/data/data/{}/shared_prefs/{}".format(
+                self.chrome_package, self.preferences_xml
+            ),
+            as_root=True,
```

Comment:
Since this is now relies on the target being rooted, you must check for that during `initialize` and raise a `WorkloadError`, so that the run aborts early if the requirement is not met.